### DISCUSSION TITLE
Option to additionally write to a log file (or to stderr)

### DISCUSSION THREAD

**Discussion by JayBrown** [marcransome/flog#69] [Ideas]
**Is your feature request related to a problem? Please describe.**
It would be nice to be able to write logs to a file too (in addition to writing to the unified log). For everyday macOS users, the unified log is very hard to comb through, so giving them an option to just open a regular log file would be a plus.

**Describe the solution you'd like**
Add either an `-f` or an `-r` option to `flog`, so we can additionally either write directly to a file, e.g. with 

```
flog -f "$HOME/Library/Logs/uk.co.fidgetbox.log" -l fault -s uk.co.fidgetbox -c general 'unrecoverable failure'
```

…or tell `flog` to write to standard error, so we can redirect that to a file, e.g. with

```
flog -r -l fault -s uk.co.fidgetbox -c general 'unrecoverable failure' 2>> "$HOME/Library/Logs/uk.co.fidgetbox.log"
```

**Describe alternatives you've considered**
(none come to mind)

**Additional context**
The macOS-built-in `logger` program does *not* have an option to write to a file directly, but it has an option (`-s`) to write to stderr, so the user can redirect.

**Comment by marcransome** [comment]
It's a feature that I'd considered adding, and something I may revisit in the near future. In the interim however, you can achieve the desired behaviour using `tee`. The message string is read from the standard input stream in this scenario, though it's not as neat as having an explicit option for it as you describe:

```bash
echo "unrecoverable error" | tee -a output.log | flog -l fault -s uk.co.fidgetbox -c general
```

Which sends the log message to Apple's unified logging system and also appends it to the `output.log` file.

There are likely other possible variations of this that may read better syntactically, though some may depend on the shell being using. For example, we can drop the `echo` command in favour of a here-string:

```bash
tee -a output.log <<< "unrecoverable error" | flog -l fault -s uk.co.fidgetbox -c general
```

This would be a trade-off in portability however, and would be limited to shells like `bash` that have support for here-strings.

Regardless, keep an eye on the [project board](https://github.com/users/marcransome/projects/3) or this issue page for updates, and thanks for the suggestion. 👍🏻

**Comment by marcransome** [comment]
@JayBrown Release [1.3.0](https://github.com/marcransome/flog/releases/tag/1.3.0) introduces a new `-a, --append` option to optionally append the log message to a file, in addition to sending it to the unified logging system. Any [feedback](https://github.com/marcransome/flog/discussions/68) would be appreciated.

**Comment by JayBrown** [comment]
Thank you very much! It seems to work so far… need to do more RL testing from within shell scripts.

Additionally, you could also mimic the way `logger` writes to log files, i.e. not just the log message, but additional info too, e.g.

```
Jan 19 10:24:11  uk.co.fidgetbox[72161] <fault>: foo
```

So you'd have the date & time followed by two whitespaces, then the process ID (the subsystem from `-s`), `flog`'s PID in square brackets without whitespace in front, then a whitespace and the log level (from `-l`) in angle brackets, a colon + whitespace, and then the log message proper.

You could also add the category (from `-c`), but I can't say how that should be done best. Maybe one of these:

```
Jan 19 10:24:11  uk.co.fidgetbox[72161] <fault:general>: foo

Jan 19 10:24:11  uk.co.fidgetbox[72161] <fault/general>: foo

Jan 19 10:24:11  uk.co.fidgetbox[72161] <fault|general>: foo
```

**Comment by JayBrown** [comment]
Addendum: you could also introduce `-A` for that, so `-a` would retain its current functionality. Because in some cases it *is* better to just log the message to a file as is. (As it stands now, I personally wouldn't need such a verbose logging functionality at the moment, but other users might.)

**Comment by marcransome** [comment]
I think it would be a worthy addition. 👍🏻 Mimicking  the format of `logger(1)` sounds sensible, and it should mean the logs are capable of being parsed by existing tools that support the format.
